File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubblePhysicalGroupView.java
package closer.vlllage.com.closer.handler.bubble;

import android.content.res.ColorStateList;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.models.Group;
import jp.wasabeef.picasso.transformations.RoundedCornersTransformation;

public class MapBubblePhysicalGroupView extends PoolMember {
    public View from(ViewGroup layer, MapBubble mapBubble, MapBubblePhysicalGroupClickListener onClickListener) {
        View view = LayoutInflater.from(layer.getContext()).inflate(R.layout.map_bubble_physical_group, layer, false);

        view.findViewById(R.id.click).setOnClickListener(v -> onClickListener.onPhysicalGroupClick(mapBubble));
        update(view, mapBubble);

        return view;
    }

    public void update(View view, MapBubble mapBubble) {
        int margin = $(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.pad);
        int size = $(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.physicalGroupIcon);
        ImageView photo = view.findViewById(R.id.photo);
        if (mapBubble.getTag() != null & mapBubble.getTag() instanceof Group) {
            Group group = (Group) mapBubble.getTag();
            if (group.getPhoto() != null) {
                margin /= 4;
                size = $(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.physicalGroupPhoto);
                photo.setColorFilter(null);
                photo.setImageTintList(ColorStateList.valueOf($(ResourcesHandler.class).getResources().getColor(android.R.color.transparent)));
                Picasso.get()
                        .load(group.getPhoto())
                        .fit()
                        .transform(new RoundedCornersTransformation($(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.physicalGroupCorners), 0))
                        .into(photo);
            } else {
                photo.setImageResource(R.drawable.ic_wifi_black_24dp);
                photo.setImageTintList(ColorStateList.valueOf($(ResourcesHandler.class).getResources().getColor(android.R.color.white)));
            }
        }

        ((ViewGroup.MarginLayoutParams) photo.getLayoutParams()).setMargins(margin, margin, margin, margin);
        ((ViewGroup.MarginLayoutParams) photo.getLayoutParams()).height = size;
        ((ViewGroup.MarginLayoutParams) photo.getLayoutParams()).width= size;
    }

    public interface MapBubblePhysicalGroupClickListener {
        void onPhysicalGroupClick(MapBubble mapBubble);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubblePhysicalGroupView.kt
package closer.vlllage.com.closer.handler.bubble

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.ViewGroup
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.databinding.MapBubblePhysicalGroupBinding
import closer.vlllage.com.closer.handler.group.PhysicalGroupHandler
import closer.vlllage.com.closer.handler.helpers.DisposableHandler
import closer.vlllage.com.closer.handler.helpers.ImageHandler
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler
import closer.vlllage.com.closer.handler.phone.NameHandler
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.store.models.Phone
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions
import com.queatz.on.On
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers

class MapBubblePhysicalGroupView constructor(private val on: On) {

    fun from(layer: ViewGroup, mapBubble: MapBubble, onClickListener: MapBubblePhysicalGroupClickListener): MapBubblePhysicalGroupBinding {
        val binding = MapBubblePhysicalGroupBinding.inflate(LayoutInflater.from(layer.context), layer, false)

        binding.click.setOnClickListener { onClickListener.invoke(mapBubble) }
        update(binding, mapBubble)

        return binding
    }

    fun update(binding: MapBubblePhysicalGroupBinding, mapBubble: MapBubble) {
        var margin = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.pad)
        var size = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.physicalGroupIcon)
        if (mapBubble.tag != null) {
            when (mapBubble.tag) {
                is Group -> {
                    if (!mapBubble.inProxy)
                        binding.click.setBackgroundResource(R.drawable.clickable_purple)

                    on<PhysicalGroupHandler>().physicalGroupName(mapBubble.tag as Group)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe({
                                if (binding.name.isAttachedToWindow) {
                                    binding.name.text = it
                                }
                            }, {}).also {
                                on<DisposableHandler>().add(it)
                            }
                }
                is Phone -> {
                    if (!mapBubble.inProxy)
                        binding.click.setBackgroundResource(R.drawable.clickable_white_rounded)

                    val status = (mapBubble.tag as Phone).status
                    binding.name.text = if (status.isNullOrBlank()) on<NameHandler>().getName((mapBubble.tag as Phone)) else "\"$status\""
                }
            }

            val photoUrl = when (mapBubble.tag) {
                is Group -> (mapBubble.tag as Group).photo
                is Phone -> (mapBubble.tag as Phone).photo
                else -> null
            }

            if (photoUrl != null) {
                margin /= 4
                size = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.physicalGroupPhoto)
                binding.photo.colorFilter = null
                binding.photo.imageTintList = ColorStateList.valueOf(on<ResourcesHandler>().resources.getColor(android.R.color.transparent))
                on<ImageHandler>().get()
                        .load("$photoUrl?s=128")
                        .apply(RequestOptions().circleCrop())
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .into(binding.photo)
            } else {
                binding.photo.setImageResource(if (mapBubble.tag is Phone) R.drawable.ic_person_black_24dp else R.drawable.ic_chat_black_24dp)
                binding.photo.imageTintList = if (mapBubble.tag is Phone) null else ColorStateList.valueOf(on<ResourcesHandler>().resources.getColor(android.R.color.white))
            }
        }

        (binding.photo.layoutParams as ViewGroup.MarginLayoutParams).setMargins(margin, margin, margin, margin)
        (binding.photo.layoutParams as ViewGroup.MarginLayoutParams).height = size
        (binding.photo.layoutParams as ViewGroup.MarginLayoutParams).width = size
    }

}

typealias MapBubblePhysicalGroupClickListener = (mapBubble: MapBubble) -> Unit
